File: interviews/leetcode/binary-tree-dfs/872. Leaf-Similar Trees/solution_recursive.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:

        # Get the leaf nodes for both trees
        leaves1 = getLeaves(root1)
        leaves2 = getLeaves(root2)

        # Compare the leaf nodes
        return leaves1 == leaves2

        # Helper function to collect the leaf nodes in a tree
def getLeaves(root):
    leaves = []
    if not root:
        return leaves
    # If it's a leaf node, add it to the list
    if not root.left and not root.right:
        leaves.append(root.val)
    # Recurse for left and right children
    leaves.extend(getLeaves(root.left))
    leaves.extend(getLeaves(root.right))
    return leaves


# Helper function to create a binary tree from a list (for testing purposes)
def create_binary_tree(lst):
    if not lst:
        return None
    root = TreeNode(lst[0])
    queue = [root]
    i = 1
    while i < len(lst):
        node = queue.pop(0)
        if lst[i] is not None:
            node.left = TreeNode(lst[i])
            queue.append(node.left)
        i += 1
        if i < len(lst) and lst[i] is not None:
            node.right = TreeNode(lst[i])
            queue.append(node.right)
        i += 1
    return root

# Test the solution
# if __name__ == "__main__":
#     solution = Solution()
#     root1 = create_binary_tree([3,5,1,6,2,9,8,None,None,7,4])
#     root2 = create_binary_tree([3,5,1,6,7,4,2,None,None,None,None,9,8])
#     result = solution.leafSimilar(root1, root2)
#     print(result)  # Output: True

File: interviews/leetcode/leetcode-75/binary-tree-dfs/872. Leaf-Similar Trees/solution_recursive.go
package main

import "fmt"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Helper function to collect the leaf nodes of a binary tree
func getLeaves(root *TreeNode) []int {
	var leaves []int
	if root == nil {
		return leaves
	}

	// If the current node is a leaf, add it to the leaves list
	if root.Left == nil && root.Right == nil {
		leaves = append(leaves, root.Val)
	}

	// Recurse for left and right subtrees
	leaves = append(leaves, getLeaves(root.Left)...)
	leaves = append(leaves, getLeaves(root.Right)...)

	return leaves
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	// Get the leaf nodes for both trees
	leaves1 := getLeaves(root1)
	leaves2 := getLeaves(root2)

	// Compare the leaf nodes of both trees
	if len(leaves1) != len(leaves2) {
		return false
	}
	for i := 0; i < len(leaves1); i++ {
		if leaves1[i] != leaves2[i] {
			return false
		}
	}
	return true
}

// Helper function to create a binary tree from a slice (for testing purposes)
func createBinaryTree(lst []interface{}) *TreeNode {
	if len(lst) == 0 || lst[0] == nil {
		return nil
	}
	root := &TreeNode{Val: lst[0].(int)}
	queue := []*TreeNode{root}
	i := 1
	for i < len(lst) {
		node := queue[0]
		queue = queue[1:]
		if lst[i] != nil {
			node.Left = &TreeNode{Val: lst[i].(int)}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(lst) && lst[i] != nil {
			node.Right = &TreeNode{Val: lst[i].(int)}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

func main() {
	// Test the solution
	root1 := createBinaryTree([]interface{}{3, 5, 1, 6, 2, 9, 8, nil, nil, 7, 4})
	root2 := createBinaryTree([]interface{}{3, 5, 1, 6, 7, 4, 2, nil, nil, nil, nil, 9, 8})

	result := leafSimilar(root1, root2)
	fmt.Println(result) // Output: true
}
